### ISSUE TITLE
Transition functions have no context access to their host object

### ISSUE THREAD

**Issue by samselikoff** [ember-animation/ember-animated#81]
I believe this is by design, but I want to make sure my understanding is correct.

If I have a `transition` function on a component

```js
export default Component.extend({

  foo: 'bar',

  transition: function*({ keptSprites }) {
    console.log(this);
    keptSprites.forEach(sprite => {
    ...
```

there's no way I can access any component state (such as `foo`) from within `transition` - correct? If so, could you explain why?

**Comment by samselikoff** [comment]
(If I had to take a stab, my best guess is so transitions are pure and thus can be reused.)

My use case at the moment (for which I'm trying to think through a better solution) is clicking an item in a list. Doing that reshuffles the list, so all sprites are `keptSprites`, but I want the `clickedItem` to be animated in a slightly different way from the others. I'm having trouble figuring out how to do that.

**Comment by ef4** [comment]
Right, they are functions, not methods. You can manually bind them if you want to, like any function.

```js
export default Component.extend({
  init() {
    this._super.init();
    this.transition = this.transition.bind(this);
  },

  foo: 'bar',

  * transition({ keptSprites }) {
    console.log(this);
    keptSprites.forEach(sprite => {
```

**Comment by samselikoff** [comment]
I see... is that considered a bad practice?

**Comment by samselikoff** [comment]
I might still be "thinking in D3" but I think my higher-level question is how to "link" Sprites with the objects/data that's driving them.

Simple demo I'm hacking on:

![Kapture 2019-03-15 at 7 18 30](https://user-images.githubusercontent.com/2922250/54437760-948fe900-46f2-11e9-8a9e-13d8092b3983.gif)

Clicking an item changes its `position` property and the list resorts. What if I want to do something "special" to the clicked item? Basically if I bound my transition function in init, maybe I could use some `data-item-id` attr on the dom element to look up the correct item from the `this.items` array, and use that to affect the animation. But it almost feels like there should be a direct way to "bind" some data to a Sprite.

**Comment by ef4** [comment]
I don't think it's bad.

We could choose to introduce `sprite.model`. There _is_ a user-provided model backing every sprite, it's just not exposed in the API today.

**Comment by samselikoff** [comment]
Is it the `predicate`, the thing that is changing that is triggering a Sprite to animate?

**Comment by ef4** [comment]
Yes, it's the value that's yielded into the block. For animated-if it's just the value of the predicate, but for a-value or a-each it's the value of the actual item.
